rules/custom_types_test.go

```diff
@@ -0,0 +1,206 @@
+package rules
+
+import (
+	"testing"
+
+	"github.com/ghodss/yaml"
+)
+
+func TestRuleGroupUnmarshalJSON(t *testing.T) {
+	for _, testCase := range []struct {
+		name string
+		raw  []byte
+		out  RuleGroup
+		err  bool
+	}{
+		{
+			name: "almost empty",
+			raw:  []byte("{}"),
+		},
+		{
+			name: "one recording rule",
+			raw: []byte(`name: foo
+interval: 5s
+rules:
+- record: bar
+  expr: vector(1)`),
+			out: RuleGroup{
+				Name:     "foo",
+				Interval: "5s",
+				Rules: []interface{}{
+					RecordingRule{
+						Record: "bar",
+						Expr:   "vector(1)",
+						Labels: RecordingRule_Labels{AdditionalProperties: make(map[string]string)},
+					},
+				},
+			},
+		},
+		{
+			name: "one alerting rule",
+			raw: []byte(`name: foo
+interval: 5s
+rules:
+- alert: HighRequestLatency
+  expr: job:request_latency_seconds:mean5m{job="myjob"} > 0.5
+  for: 10m`),
+			out: RuleGroup{
+				Name:     "foo",
+				Interval: "5s",
+				Rules: []interface{}{
+					AlertingRule{
+						Alert:       "HighRequestLatency",
+						Expr:        `job:request_latency_seconds:mean5m{job="myjob"} > 0.5`,
+						For:         "10m",
+						Annotations: AlertingRule_Annotations{AdditionalProperties: make(map[string]string)},
+						Labels:      AlertingRule_Labels{AdditionalProperties: make(map[string]string)},
+					},
+				},
+			},
+		},
+		{
+			name: "one of each",
+			raw: []byte(`name: foo
+interval: 5s
+rules:
+- record: bar
+  expr: vector(1)
+- alert: HighRequestLatency
+  expr: job:request_latency_seconds:mean5m{job="myjob"} > 0.5
+  for: 10m`),
+			out: RuleGroup{
+				Name:     "foo",
+				Interval: "5s",
+				Rules: []interface{}{
+					RecordingRule{
```

Comment:
so we can still get both recording and alerting rules in the same request, but we wouldn't like to have a rule that has both `alert` and `record` properties, right? is that why `anyOf` was changed to `oneOf`? https://github.com/observatorium/api/pull/191/commits/d08124476f3ec0716dadafe6bc8005767439be51#diff-d523f469970611999b9968425756e5cc8c840626d7e592cf514492c52215c9e6L81
Just wanted to make sure I understand since here we expect one `RuleGroup` with both alerting/recording rules

Reply:
That's correct. If it was anyOf, then a single rule would be allowed to have both `alert` and `record` fields defined, which is not what we want. https://swagger.io/docs/specification/data-models/oneof-anyof-allof-not/ that's my reading at least. The section `Difference Between anyOf and oneOf` seems to suggest that

Reply:
oki, this makes sense :+1: